Improve site packages directory retrieval across platforms #11949

aj-he opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->
### Types of changes
<!--- Put an `x` in the boxes that apply  -->
-   [x] Bugfix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Description of Change
<!--- Include a brief description of this Pull Request and reasoning behind it. -->
Hardens the site-packages directory path lookup for Windows users. Arches app frontend assets cannot be found on Windows otherwise.

### Issues Solved
<!--- If this Pull Request solves any issues, list them here, and mark the ticket "In Review" in the pipeline project -->
Closes #11949

### Checklist
<!--- Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.  -->
-   I targeted one of these branches:
    - [ ] dev/8.0.x (under development): features, bugfixes not covered below
    - [x] dev/7.6.x (main support): regressions, crashing bugs, security issues, major bugs in new features
    - [ ] dev/6.2.x (extended support): major security issues, data loss issues
-   [x] I added a changelog in arches/releases
-   [ ] I submitted a PR to arches-docs (if appropriate)
-   [x] Unit tests pass locally with my changes
-   [x] I added tests that prove my fix is effective or that my feature works
-   [ ] My test fails on the target branch

#### Accessibility Checklist
<!-- If your changes impacted the following areas, mark the appropriate columns. -->
[Developer Guide](https://arches.readthedocs.io/en/stable/developing/advanced/accessibility/)

| Topic            | Changed | Retested |
| ---------------- | ------- | -------- |
| Color contrast   |         |          |
| Form fields      |         |          |
| Headings         |         |          |
| Links            |         |          |
| Keyboard         |         |          |
| Responsive Design|         |          |
| HTML validation  |         |          |
| Screen reader    |         |          |


#### Ticket Background
*   Sponsored by: Historic England
*   Found by: @aj-he 
*   Tested by: @aj-he 
*   Designed by: @aj-he 

### Further comments

<!--- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

aj-he committed:
improve site packages directory retrieval across platforms #11949

aj-he committed:
Update release notes for Arches 7.6.10 #11949

aj-he committed:
nit black #11949

jacobtylerwalls commented on arches/settings_utils.py:
```diff
@@ -152,12 +152,22 @@ def generate_frontend_configuration():
             "ARCHES_APPLICATIONS": arches_app_names,
             "ARCHES_APPLICATIONS_PATHS": path_lookup,
-            "SITE_PACKAGES_DIRECTORY": site.getsitepackages()[0],
+            "SITE_PACKAGES_DIRECTORY": site_packages_dir,
```
Thanks for finding this bug. I wonder if we can just do:
```suggestion
            "SITE_PACKAGES_DIRECTORY": sysconfig.get_path("purelib"),
```

aj-he replied:
@jacobtylerwalls - I'll take a look at this to see what happens on Windows/Linux etc.

aj-he replied:
@jacobtylerwalls - yep... worked a charm and a lot more concise.

aj-he committed:
Refactor site packages directory retrieval using sysconfig #11949

aj-he committed:
Merge branch 'dev/7.6.x' into 11949_windows_frontend_config_bug

aj-he committed:
Add unit tests for frontend configuration generation #11949
- only added generate_frontend_configuration as that is what has been refactored

aj-he commented:
@jacobtylerwalls - you ok to review this for me? I've added a small test for the fix as it was failing test coverage. I did see that the function has been moved in 8.x but hopefully not too tricky to merge?

jacobtylerwalls commented:
Certainly, I'll take a look likely tomorrow. Only to suggest some simplifications to the tests.

jacobtylerwalls commented on releases/7.6.10.md:
```diff
@@ -3,6 +3,7 @@
 ### Bug Fixes and Enhancements
 
 - Fix regression in original fix (#11748) for slow reports when used on systems with many related resources [#11959](https://github.com/archesproject/arches/pull/11959)
+- Fixes frontent build issue on Windows [#11949](https://github.com/archesproject/arches/pull/11949)
```
```suggestion
- Fixes frontend build issue on Windows [#11949](https://github.com/archesproject/arches/pull/11949)
```

aj-he replied:
done

jacobtylerwalls commented on releases/7.6.10.md:
````diff
@@ -28,7 +29,24 @@ JavaScript:
     ```
 
-3. If you are running Arches on Apache, restart your server:
+3.  If you are upgrading from version 7.6.9, you'll need to rebuild your frontend bundle:
````
Is this strictly true? The frontend changes are addressing edge cases that caused bundles to not build at all, so if you've got a bundle, wouldn't you be all set?

jacobtylerwalls replied:
Ah, I take it the arches apps were omitted, perhaps silently? Then let's leave this in. 👍

jacobtylerwalls commented on tests/utils/test_settings_utils.py:
```diff
@@ -0,0 +1,164 @@
+        """
+        Test frontend configuration generation when running core arches directly.
+        """
+        # Mock settings
+        mock_settings = mock.MagicMock()
+        mock_settings.APP_ROOT = "/arches/app"
+        mock_settings.ROOT_DIR = "/arches"
+        mock_settings.PUBLIC_SERVER_ADDRESS = "http://localhost:8000"
+        mock_settings.STATIC_URL = "/static/"
```
I'd like to stick to the existing pattern for overriding settings during tests via the `@override_settings` decorator and thereby remove as much mocking as possible.
```suggestion
```

aj-he replied:
applied

jacobtylerwalls commented on tests/utils/test_settings_utils.py:
```diff
@@ -0,0 +1,164 @@
+        mock_settings.ROOT_DIR = "/arches"
+        mock_settings.PUBLIC_SERVER_ADDRESS = "http://localhost:8000"
+        mock_settings.STATIC_URL = "/static/"
+        mock_settings.WEBPACK_DEVELOPMENT_SERVER_PORT = 8080  # Add this missing setting
```
This setting isn't missing from the test settings.

aj-he replied:
done

jacobtylerwalls commented on tests/utils/test_settings_utils.py:
```diff
@@ -0,0 +1,164 @@
+# python manage.py test tests.utils.test_settings_utils --settings="tests.test_settings"
+
+
+class TestFrontendConfigurationGeneration(TestCase):
```
Thanks for adding this. Now that we have a test in place, let's keep ourselves accountable by no longer excluding it from the coverage calculation. Would you mind adjusting the `settings*.py` to `*settings.py` or similar?

aj-he replied:
@jacobtylerwalls - I'm slightly confused by this. When I submitted my initial change, the coverage check failed the PR despite me only changing the `settings_utils.py` and `release/7.6.x.md` files. The above note suggests that `settings_utils.py` would have been excluded? This is the only reason I added the test!!

aj-he replied:
... also, if I include `settings_utils.py` in the coverage at this point, am I going to have to write tests for the other functions in the module for this to be accepted?

jacobtylerwalls replied:
Oh I certainly wouldn't ask you to volunteer for testing a bunch of unrelated stuff, no worries.

>  the coverage check failed the PR despite me only changing the settings_utils.py

I can't speak to why that comparison failed, but I wonder if there was a slight drift between your branch and upstream dev/7.6.x

If you go to the actions tab and look at the coverage reports generated, you'll see there's no arches/settings_utils.py being measured

jacobtylerwalls replied:
The coverage check has to be taken with a grain of salt because of the "falling denominator" problem -- it's always acceptable to remove tested code.

aj-he replied:
thanks for reviewing and sorting out @jacobtylerwalls.

jacobtylerwalls commented on tests/utils/test_settings_utils.py:
```diff
@@ -0,0 +1,164 @@
+        Test error handling in frontend configuration generation.
+        """
+        # Force an exception by removing a required setting
+        with mock.patch("arches.settings_utils.settings") as mock_settings:
```
Same as above, you can use `@override_settings` or a variant like `self.modify_settings()`.

aj-he replied:
done

jacobtylerwalls commented on tests/utils/test_settings_utils.py:
```diff
@@ -0,0 +1,164 @@
+        with mock.patch("arches.settings_utils.settings") as mock_settings:
+            delattr(mock_settings, "APP_ROOT")
+
+            # This should raise an AttributeError
+            with self.assertRaises(Exception):
```
```suggestion
            with self.assertRaises(AttributeError):
```

jacobtylerwalls commented on tests/utils/test_settings_utils.py:
```diff
@@ -0,0 +1,164 @@
+            {"@/arches/*", "@/test_app/*", "*"},
+        )
+
+    @mock.patch("arches.settings_utils.sys.stderr.write")
```
We can avoid mocking with the more common:
```py
with captured_stderr() as stderr:
   # write to stderr
   self.assertIn('foo', stderr.getvalue())
```

aj-he replied:
done

jacobtylerwalls commented on tests/utils/test_settings_utils.py:
```diff
@@ -0,0 +1,164 @@
+        mock_open.assert_any_call("/arches/../.tsconfig-paths.json", "w")
+
+        # Check content of first file
+        expected_settings_data = {
```
I wouldn't be afraid of removing much of the mocking and writing this file to a tempdir, but I leave that to you.

aj-he replied:
I've stripped most out now

jacobtylerwalls reviewed (changes requested):
Thanks for the updates, just a few suggestions in the tests, let me know what you think. Mostly I just want to avoid having the tests look forbidding to the next developer :-)

aj-he committed:
removed mock in favor of existing patterns #11949

aj-he committed:
Merge branch 'dev/7.6.x' into 11949_windows_frontend_config_bug

aj-he committed:
nit spelling #11949

jacobtylerwalls committed:
Remove APP_ROOT from test_settings re #11949

jacobtylerwalls committed:
Restore site import re #11949
